sklearn/cluster/_kmeans.py

```diff
@@ -1410,102 +1607,146 @@ def fit(self, X, y=None, sample_weight=None):
-            ):
-                best_labels = labels
-                best_centers = centers
-                best_inertia = inertia
-                best_n_iter = n_iter_
+    def fit_transform(self, X, y=None, sample_weight=None):
+        """Compute clustering and transform X to cluster-distance space.
 
-        if not sp.issparse(X):
-            if not self.copy_x:
-                X += X_mean
-            best_centers += X_mean
+        Equivalent to fit(X).transform(X), but more efficiently implemented.
 
-        distinct_clusters = len(set(best_labels))
-        if distinct_clusters < self.n_clusters:
-            warnings.warn(
-                "Number of distinct clusters ({}) found smaller than "
-                "n_clusters ({}). Possibly due to duplicate points "
-                "in X.".format(distinct_clusters, self.n_clusters),
-                ConvergenceWarning,
-                stacklevel=2,
-            )
+        Parameters
+        ----------
+        X : {array-like, sparse matrix} of shape (n_samples, n_features)
+            New data to transform.
 
-        self.cluster_centers_ = best_centers
-        self._n_features_out = self.cluster_centers_.shape[0]
-        self.labels_ = best_labels
-        self.inertia_ = best_inertia
-        self.n_iter_ = best_n_iter
-        return self
+        y : Ignored
+            Not used, present here for API consistency by convention.
+
+        sample_weight : array-like of shape (n_samples,), default=None
+            The weights for each observation in X. If None, all observations
+            are assigned equal weight.
+
+        Returns
+        -------
+        X_new : ndarray of shape (n_samples, n_clusters)
+            X transformed in the new space.
+        """
+        # XXX pre_transform() is not called because fit() calls pre_fit()
+        self.fit(X, sample_weight=sample_weight)
+        engine = self._get_engine(X)
+        return self._transform(X, engine)
+
+    def transform(self, X):
+        """Transform X to a cluster-distance space.
+
+        In the new space, each dimension is the distance to the cluster
+        centers. Note that even if X is sparse, the array returned by
+        `transform` will typically be dense.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix} of shape (n_samples, n_features)
+            New data to transform.
+
+        Returns
+        -------
+        X_new : ndarray of shape (n_samples, n_clusters)
+            X transformed in the new space.
+        """
+        check_is_fitted(self)
+        engine = self._get_engine(X)
+        if hasattr(engine, "pre_transform"):
+            X = engine.pre_transform(X)
+        return self._transform(X, engine)
+
+    def _transform(self, X, engine):
+        """Guts of transform method; no input validation."""
+        X_ = engine.transform(X)
+        if hasattr(engine, "post_transform"):
+            engine.post_transform(X_)
```

Comment:
Currently `pre_transform` is done in `transform` while `post_transform` is done in `_transform`.

Hence, are we sure that `transform` is the only function which calls `_transform` so that `pre_transform` will always be called? Must it always be called? Similarly `post_transform` is currently always called but must it be?